alf/networks/normalizing_flow_networks.py

```diff
@@ -0,0 +1,408 @@
+        many transformation layers.
+
+        Args:
+            conditional_inputs: an optional nested conditional inputs that
+                condition the mapping :math:`x \rightarrow y`.
+
+        Returns:
+            torch.td.Transform: an invertible transform
+        """
+        if not self._use_transform_cache:
+            return self._make_invertible_transform(conditional_inputs)
+        old_t, old_z = self._cached_transform
+        if conditional_inputs is old_z and old_t is not None:
+            return old_t
+        t = self._make_invertible_transform(conditional_inputs)
+        self._cached_transform = (t, conditional_inputs)
+        return t
+
+    def _make_invertible_transform(
+            self, conditional_inputs: alf.nest.NestedTensor = None):
+        raise NotImplementedError()
+
+    def forward(self,
+                xz: Union[torch.Tensor, Tuple[torch.Tensor, alf.nest.
+                                              NestedTensor]],
+                state: alf.nest.NestedTensor = ()):
+        """When we have no conditional input for forward: ``y=self.forward(x)``.
+        Otherwise ``y=self.forward((x,z))`` where ``z`` is the conditional input.
+
+        Args:
+            xz: the input can be either an unnested tensor ``x`` or a tuple of
+                an unnested tensor and a nested tensor ``(x, z)``. In both cases,
+                currently we only support ``x`` to be a rank-1 tensor. ``z`` is
+                an optional conditional input that conditions the normalizing
+                flow mapping from ``x`` to ``y``.
+            state: should be an empty tuple
+        """
+        if self._conditional_inputs:
+            x, z = xz
+        else:
+            x, z = xz, None
+        transform = self.make_invertible_transform(z)
+        return transform(x), ()
+
+    def inverse(self,
+                yz: Union[torch.Tensor, Tuple[torch.Tensor, alf.nest.
+                                              NestedTensor]],
+                state: alf.nest.NestedTensor = ()):
+        """When we have no conditional input for forward: ``x=self.inverse(y)``.
+        Otherwise ``x=self.inverse((y,z))`` where ``z`` is the conditional input.
+
+        Args:
+            yz: the input can be either an unnested tensor ``y`` or a tuple of
+                an unnested tensor and a nested tensor ``(y, z)``. In both cases,
+                currently we only support ``y`` to be a rank-1 tensor. ``z`` is
+                an optional conditional input that conditions the normalizing
+                flow inverse mapping from ``y`` to ``x``.
+            state: should be an empty tuple
+        """
+        if self._conditional_inputs:
+            y, z = yz
+        else:
+            y, z = yz, None
+        transform = self.make_invertible_transform(z)
+        return transform.inv(y), ()
+
+
+@alf.configurable
+class realNVPNetwork(NormalizingFlowNetwork):
```

Comment:
real => Real ?

Reply:
> real => Real ?

'realNVP' is what the paper uses to call their method.